packages/labs/src/multi-select/multiSelect.tsx

```diff
@@ -0,0 +1,299 @@
+                    </ul>
+                </div>
+            </Popover>
+        );
+    }
+
+    private renderItems({ activeItem, filteredItems, handleItemSelect }: IQueryListRendererProps<T>) {
+        const { itemRenderer, noResults } = this.props;
+
+        if (filteredItems.length === 0) {
+            return noResults;
+        }
+
+        return filteredItems.map((item, index) => itemRenderer({
+            index,
+            item,
+            handleClick: (e) => handleItemSelect(item, e),
+            isActive: item === activeItem,
+            isSelected: this.isItemSelected(item),
+        }));
+    }
+
+    private handleQueryChange = (event: React.FormEvent<HTMLInputElement>) => {
+        const query = event.currentTarget.value;
+        let nextState = { query };
+
+        if (query.length === 0 && this.props.openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+
+        this.setState(nextState);
+    }
+
+    private getSelectedItemIndex(item: T) {
+        return this.state.selectedItems.indexOf(item);
+    }
+
+    private isItemSelected(item: T) {
+        return this.getSelectedItemIndex(item) !== -1;
+    }
+
+    private selectItem(item: T) {
+        const { openOnKeyDown, resetOnSelect } = this.props;
+        let nextState = { selectedItems: [...this.state.selectedItems, item] };
+
+        if (openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+        if (resetOnSelect) {
+            nextState = { ...nextState, query: "" };
+        }
+
+        this.setState(nextState);
+    }
+
+    private deselectItem(index: number) {
+        this.setState({ selectedItems: this.state.selectedItems.filter((_, i) => i !== index) });
+    }
+
+    private handleTagRemove = (_item: string, index: number) => {
+        this.deselectItem(index);
+
+        Utils.safeInvoke(this.props.tagInputProps.onRemove, _item, index);
+    }
+
+    private handleItemSelect = (item: T, _event: React.SyntheticEvent<HTMLElement>) => {
+        // check whether the query string is in the item list
+        if (item !== undefined && this.state.activeItem !== undefined) {
+            if (!this.isItemSelected(item)) {
+                this.selectItem(item);
+            } else {
+                this.deselectItem(this.getSelectedItemIndex(item));
+            }
+        }
+        this.input.focus();
+
+        Utils.safeInvoke(this.props.onItemSelect, item, _event);
+    }
+
+    private handlePopoverInteraction = (nextOpenState: boolean) => requestAnimationFrame(() => {
+        // deferring to rAF to get properly updated activeElement
+        const { popoverProps = {} } = this.props;
+
+        if (this.input != null && this.input !== document.activeElement) {
+            // the input is no longer focused so we can close the popover
+            let nextState = { isOpen: false };
+
+            if (this.props.resetOnSelect) {
```

Comment:
hmmm this logic happens at least three times. i'm sure we can come up with a helpful util to handle it only once, but i don't have the mental space for that right now 🌴